llvm/test/CodeGen/NVPTX/sm-version.ll

```
@@ -56,6 +56,64 @@
 ; RUN: llc < %s -mtriple=nvptx64 -mcpu=sm_121 | FileCheck %s --check-prefix=SM121
 ; RUN: llc < %s -mtriple=nvptx64 -mcpu=sm_121a | FileCheck %s --check-prefix=SM121a
 
+; RUN: %if ptxas %{ llc < %s -mtriple=nvptx -mcpu=sm_20 | %ptxas-verify %}
```

All of these need to be careful about ptxas versions.
We do not have control over which ptxas version the user may use, and that particular version may support only a subset of the GPUs.

So, for sm_20, it would have to be `%if ptxas && !ptxas-10.0`. 
You will also need to adjust `-arch` flag appropriately for the `-ptxas-verify`, otherwise it will default to sm_60 (I think).

Does this test actually succeed in the current form when you run it with PTX enabled?

I suspect that it does not -- I would expect ptxas-verify to complain about the GPU for all GPUs newer than sm_60. It will also silently accept pre-sm_60 GPUs, and will happily compile them for sm_60, but it's not clear whether it's a test bug (we're not verifying compilation for the requested GPU) or a feature (we did verify that ptxas can successfully parse PTX generated by llc, and whether ptxas can actually target the GPU the PTX was intended for not LLVM's concern)

---

Yeah okay, Understood.
From PTX ISA: http://sw-mobile-docs/CUDA/GPGPU/parallel-thread-execution/index.html#ptx-module-directives-target
Target string sm_20 introduced in PTX ISA version 2.0.
Target string sm_30 introduced in PTX ISA version 3.0.
Target string sm_35 introduced in PTX ISA version 3.1.
Target strings sm_32 and sm_50 introduced in PTX ISA version 4.0.
Target strings sm_37 and sm_52 introduced in PTX ISA version 4.1.  and so on.....
And sm_20 will be supported with all PTX ISA version 2.0+. So In the line for sm_20, can we have like below:
; RUN: %if ptxas && ptxas-2.0+ %{ llc < %s -mtriple=nvptx -mcpu=sm_20 | %ptxas-verify -arch=sm_20 %}
instead of "%if ptxas && !ptxas-10.0"
Let me know what you think..
Also need some modifications in the PTX ISA versions wrt to the target based on the PTX ISA page.

---

Not quite. 

`ptxas-X.Y` means "ptxas version X.Y or newer" which usually aligns with "CUDA SDK version X.Y". You need to find the range of CUDA versions that support the SM you need (https://en.wikipedia.org/wiki/CUDA#GPUs_supported) and the range of versions that support the PTX version you want (https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#release-notes-ptx-release-history)
Then constrain ptxas version to the union of those CUDA version ranges.

---

Thanks Artem , this format indeed catches the range of CUDA version that support the SM needed. (Something like the below)
```
; RUN: %if ptxas && !ptxas-10.0 %{ llc < %s -mtriple=nvptx -mcpu=sm_20 | %ptxas-verify -arch=sm_20 %}
; RUN: %if ptxas && !ptxas-11.0 %{ llc < %s -mtriple=nvptx -mcpu=sm_21 | %ptxas-verify -arch=sm_21 %}
.<rest of it>..
```